Python/pystate.c

```diff
@@ -829,11 +829,16 @@ new_threadstate(PyInterpreterState *interp)
         // Every valid interpreter must have at least one thread.
         assert(id > 1);
         assert(old_head->prev == NULL);
-
+        // Unlock before allocating memory.
+        HEAD_UNLOCK(runtime);
```

Comment:
If possible, I prefer to move the allocation at the beginning as @tom-pytel is suggesting because is much easier to reason about. Here you are interrupting the critical section in the middle which is potentially incorrect depending on what tracemalloc does (currently it done not do anything fancy like spawning threads but you are implicitly relyi ng on that contract).

Reply:
If you concerned about this I can change it to always allocate memory and free it if unused but I would like  to avoid it unless necessary.

Reply:
I see what you mean and is great that you want to ensure maximum performance but I personally think that is premature optimization. Creating thread states is not a performance-critical part of the code and this problem with the allocation being unused only happens if tracemalloc is in use, which impacts performance by a ton already, so this is a negligible advantage at the cost of making reasoning about the correctness of the code harder.

Reply:
>  Creating thread states is not a performance-critical part of the code and this problem with the allocation being unused only happens if tracemalloc is in use,

Just a clarification that the allocation can be unused even without tracemalloc if `interp->threads.head is NULL` which is the common case of this function. But I see your point and will change. Thanks for your feedback.